### PR TITLE
feat: Support variadic arguments in the middle of a function signature

### PR DESCRIPTION
**Pull request by reneesoika**
### Description 
Follow-up to #9361. This PR adds support for a variadic argument anywhere in the function signature. There can still be only one variadic argument in the function signature.

This change is being made because we currently cannot support variadic UDAFs with initial arguments. Allowing a variadic argument in the middle removes this restriction.

The main changes to enable this feature are in `UdfIndex`. The tree of parameter types is now based on pairs of parameters, with one parameter from the left half of the parameter list and one parameter from the right half. This structure ensures the loop for variadic arguments is always at the bottom of the tree, eliminating most edge cases. There is also a new rule to determine function precedence:

> If two methods exist that match given the above rules, return the method with the variadic argument in the later position.

The number of nodes generated is quadratic with respect to the number of parameters in the function (only for variadic functions), but we only generate the tree once at startup. The number of nodes generated is not excessive unless there is a grossly absurd number of arguments ([see graph](https://www.desmos.com/calculator/fbnikgl7ow)). Currently, the only way to have a function with this many arguments is to have many initial arguments, as Java supports up to 255 arguments for a method. Resolving a function still takes linear time.

How `UdfIndex` now works is documented more thoroughly in comments. Other changes are removing exceptions thrown for var args in the middle or tests.

### Testing done 
Added unit tests and QTTs for var args in the middle.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-common/src/main/java/io/confluent/ksql/function/UdfIndex.java [hunk 1]
```diff
@@ -104,44 +110,116 @@
...
               + " with parameters " + function.parameters()
               + " as a function with the same name and parameter types already exists "
-              + allFunctions.get(parameters)
+              + allFunctions.get(function.parameters())
       );
     }
 
+    /* Build the tree for non-variadic functions, or build the tree that includes one variadic
+    argument for variadic functions. */
+    final Pair<Node, Integer> variadicParentAndOffset = buildTree(
```
[the remark this answers is not in the document]
Comment (reneesoika) [reply ksqldb-common/src/main/java/io/confluent/ksql/function/UdfIndex.java:125]: Spoke with Zara about this: the order in which you build the tree doesn't matter, so this is arbitrary. The existing `UdfIndex` implementation allows you to input a list for a var args parameter, and this implementation preserves that.

File: ksqldb-common/src/test/java/io/confluent/ksql/function/UdfIndexTest.java [hunk 2]
```diff
@@ -489,11 +555,31 @@ public void shouldFindVarargWithList() {
   }
 
   @Test
-  public void shouldChooseSpecificOverVarArgs() {
+  public void shouldNotCreateListWhenNotVariadic() {
```
[the remark this answers is not in the document]
Comment (reneesoika) [reply ksqldb-common/src/test/java/io/confluent/ksql/function/UdfIndexTest.java:558]: Added this test, as well as another test to check what happens if you add a variadic function and a non-variadic function that accepts a list of the same type as the vararg. Verified behavior is same as master (exception).

**Review by jzaralim — APPROVED** [review]
lgtm thanks @reneesoika

**Comment by reneesoika** [comment]
@jzaralim I noticed a minor bug that would cause the function call to be ambiguous, which I just pushed a commit to fix.

If you had these two function signatures:
```
f(A first, int second, boolean... third)
g(B first, int... second, boolean third)
```

`f` and `g` would end up on different branches, so the comparison of variadic position in `Node#update()` would never be executed, leading to an ambiguous function call exception when both match.

I moved all the precedence rule comparisons into the same comparator, which fixes the issue. In the original `UdfIndex`, I think the generics rule was separate from the comparator because two functions with different numbers of generics could never be the value of the same node. IMO, putting all the rules into the same comparator is going to be less bug-prone since the precedence rules must always be followed, even if they aren't all necessary in a particular spot.

**Pull request merged**